go/libraries/doltcore/dolttypes/dolttype.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2019 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dolttypes
+
+import (
+	"fmt"
+)
+
+// DoltKind is Dolt's version of NomsKind for Dolt-defined types built on UnderlyingArrayKind.
+type DoltKind byte
+const DoltKindLength = 1
+
+const (
+	UnknownKind DoltKind = iota
+	TimestampKind
+)
+
+type DoltType interface {
```

Comment:
I think this is probably the wrong approach. We don't want to mimic the Noms storage layer in the dolt layer, just build on top of it.

Rather than have a separate NomsKind and DoltKind everywhere, you probably want to just stick with the NomsKind constants, and mark some of them only as used by Dolt for record keeping, not actually reified as those types. Example:

        BoolKind NomsKind = iota
	FloatKind
	StringKind
        ...
        ByteArrayKind // New noms type you're implementing, fully reified in noms layer
        DoltTimestampKind // Not ever directly stored in noms, just used to identify the type stored in the ByteArrayKind in a dolt Schema
        DoltPointKind // Not ever directly stored in noms, just used to identify the type stored in the ByteArrayKind in a dolt Schema

Aaron should take a look as well.

Reply:
This would require one to add `ByteArrayKind`, `DoltTimestampKind`, and `DoltPointKind` in every place that's doing a check for the type, which isn't clear. Even for my (name changing soon) `UnderlyingArrayKind`, I'm not 100% sure I've added the type to all of the places that handles the kind at this moment, so to force that workflow seems prone to introducing bugs.